npm/packages/@socketsupply/socket/node/index.js

```diff
@@ -0,0 +1,306 @@
+// @ts-check
+import { format } from 'node:util'
+import { writeSync, fsyncSync } from 'node:fs'
+
+const AUTO_CLOSE = process.env.AUTO_CLOSE
+const MAX_MESSAGE_KB = 512 * 1024
+
+function isObject (o) {
+  return o && typeof o === 'object' && !Array.isArray(o)
+}
+
+const write = s => {
+  if (s.includes('\n')) {
+    throw new Error('invalid write()')
+  }
+
+  if (s.length > MAX_MESSAGE_KB) {
+    const len = Math.ceil(s.length / 1024)
+    process.stderr.write('WARNING: Sending large message to webview: ' + len + 'kb\n')
+    process.stderr.write('RAW MESSAGE: ' + s.slice(0, 512) + '...\n')
+  }
+
+  return new Promise(resolve =>
+    process.stdout.write(s + '\n', resolve)
+  )
+}
+
+console.log = (...args) => {
+  const s = args.map(v => format(v)).join(' ')
+  const enc = encodeURIComponent(s)
+  // fs.appendFileSync('tmp.log', s + '\n')
+  write(`ipc://stdout?value=${enc}`)
+}
+console.error = console.log
+
+process.on('exit', (exitCode) => {
+  const seq = String(ipc.nextSeq++)
+
+  let value = new URLSearchParams({
+    index: '0',
+    seq,
+    value: String(exitCode)
+  }).toString()
+
+  value = value.replace(/\+/g, '%20')
+
+  writeSync(1, `ipc://exit?${value}\n`)
+  try {
+    fsyncSync(1)
+  } catch (_) {
+    // fsync(1) can fail in github actions for reasons unclear.
+    // maybe the stdout is weird in that environment.
+  }
+})
+
+//
+// Internal IPC API
+//
+const ipc = { nextSeq: 0 }
+
+ipc.resolve = async (seq, state, value) => {
+  const method = !Number(state) ? 'resolve' : 'reject'
+  if (!ipc[seq] || !ipc[seq][method]) return
+
+  /**
+   * TODO Who handles this error ?
+   */
+  try {
+    await ipc[seq][method](value)
+  } finally {
+    delete ipc[seq]
+  }
+}
+
+ipc.request = async (cmd, opts) => {
+  const seq = ipc.nextSeq++
+  let value = ''
+
+  const promise = new Promise((resolve, reject) => {
+    ipc[seq] = {
+      resolve: resolve,
+      reject: reject
+    }
+  })
+
+  try {
+    if (typeof opts.value === 'object') {
+      opts.value = JSON.stringify(opts.value)
+    }
+
+    value = new URLSearchParams({
+      ...opts,
+      index: opts.window || '0',
+      seq,
+      value: opts.value || '0'
+    }).toString()
+
+    value = value.replace(/\+/g, '%20')
+  } catch (err) {
+    console.error(`Cannot encode request ${err.message} (${value})`)
+    return Promise.reject(err)
+  }
+
+  await write(`ipc://${cmd}?${value}`)
+  return promise
+}
+
+ipc.send = async o => {
+  try {
+    // TODO: use structuredClone instead once we are on node 17+
+    o = JSON.parse(JSON.stringify(o))
```

Comment:
You'd be surprised to know that `structuredClone` is much slower [1]

1. https://www.measurethat.net/Benchmarks/Show/18541/0/jsonstringify-vs-structuredclone#latest_results_block